src/ConnectionConfig.tsx

```diff
@@ -1,29 +1,43 @@
 import { DataSourcePluginOptionsEditorProps, onUpdateDatasourceJsonDataOption } from '@grafana/data';
 import { Alert, Field, FieldSet, RadioButtonGroup } from '@grafana/ui';
+import { BackendSrv } from '@grafana/runtime';
 
-import React from 'react';
+import React, { useState } from 'react';
 import { TEST_IDS } from 'testIds';
 import { JWTConfigEditor } from './components/JWTConfigEditor';
 import { JWTForm } from './components/JWTForm';
 import { GOOGLE_AUTH_TYPE_OPTIONS } from './constants';
 import { DataSourceOptions, DataSourceSecureJsonData, GoogleAuthType } from './types';
 
-export type ConfigEditorProps = DataSourcePluginOptionsEditorProps<DataSourceOptions, DataSourceSecureJsonData>;
+export type ConfigEditorProps = DataSourcePluginOptionsEditorProps<DataSourceOptions, DataSourceSecureJsonData> & {
+  backendSrv?: BackendSrv;
+};
 
 export const ConnectionConfig: React.FC<ConfigEditorProps> = (props: ConfigEditorProps) => {
   const { options, onOptionsChange } = props;
   const { jsonData, secureJsonFields, secureJsonData } = options;
+  const [error, setError] = useState<string | undefined>(undefined);
 
   if (!jsonData.authenticationType) {
     jsonData.authenticationType = GoogleAuthType.JWT;
   }
 
   const isJWT = jsonData.authenticationType === GoogleAuthType.JWT || jsonData.authenticationType === undefined;
 
-  const onAuthTypeChange = (authenticationType: GoogleAuthType) => {
+  const onAuthTypeChange = async (authenticationType: GoogleAuthType) => {
+    let gceDefaultProject = jsonData.gceDefaultProject;
+    try {
+      if (authenticationType === 'gce' && !gceDefaultProject) {
+        gceDefaultProject = await props.backendSrv.get(
+          `/api/datasources/${props.options.id}/resources/gceDefaultProject`
+        );
+      }
+    } catch (err) {
+      setError(err.data.message);
+    }
```

Comment:
actually, I don't know if it's the best to do this here. It's doing a backend call before the settings are saved. This means that people will get an error just by clicking the auth type switch (if they are not in GCE) and any other setting needed for the configuration won't be taken into account.

I think that the problem is that the cloud-monitoring plugin is defining this `gceDefaultProject` as part of the JsonData, but it's not (since it's not being saved here). I would remove it from there:

https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/cloud-monitoring/types.ts#L159

And replace it with Datasource private property, initializing its value in the constructor. That way we avoid this ambiguity.

Reply:
I don't think it's necessarily a problem that we're doing a backend call here before settings are saved but I see your point on the error.

I agree on the comment around the two different project names and I'll look into potentially making that change instead. Setting the project name in the datasource would also avoid any issues if an instance were to be moved to a different project.